lab_4/main.py

```diff
@@ -6,64 +6,236 @@
+                      isinstance(context, tuple)]
+        if not all(conditions) or word not in self._word_storage.storage.values() or \
+                len([w for w in context if w in self._word_storage.storage.values()]) != len(context):
+            raise ValueError
+
+        word_freq = 0
+        average_freq = 0
+        for n_gram in self._n_gram_trie.n_grams:
+            if context == n_gram[:len(context)]:
+                average_freq += 1
+                if word == n_gram[-1]:
+                    word_freq += 1
+
+        try:
+            likelihood = word_freq / average_freq
+        except ZeroDivisionError:
+            likelihood = 0.0
+        return likelihood
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+        if not isinstance(context, tuple) or \
+                len([w for w in context if w in self._word_storage.storage.values()]) != len(context):
+            raise ValueError
+        next_word = -1
+        word_freq = -1
+        for word in self._word_storage.storage.values():
+            local_freq = self._calculate_maximum_likelihood(word, context)
+            if local_freq > word_freq:
+                word_freq = local_freq
+                next_word = word
+        return next_word
 
 
 class BackOffGenerator(NGramTextGenerator):
 
     def __init__(self, word_storage: WordStorage, n_gram_trie: NGramTrie, *args):
         super().__init__(word_storage, n_gram_trie)
+        self._n_gram_tries = (n_gram_trie, *args)
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+        if not isinstance(context, tuple) or \
+                len([w for w in context if w in self._word_storage.storage.values()]) != len(context):
+            raise ValueError
+
+        for trie in self._n_gram_tries:
+            next_word = -1
+            next_word_freq = 0
+            for n_gram, freq in trie.n_gram_frequencies.items():
+                if context[:trie.size - 1] == n_gram[:trie.size - 1]:
+                    if freq > next_word_freq:
+                        next_word_freq = freq
+                        next_word = n_gram[-1]
+            if next_word != -1:
+                return next_word
+
+        uni = list(self._n_gram_trie.uni_grams.items())
+        uni.sort(key=lambda x: x[1], reverse=True)
+        next_word = uni[0][0][0]
+        return next_word
 
 
 def decode_text(storage: WordStorage, encoded_text: tuple) -> tuple:
-    pass
+    if not isinstance(storage, WordStorage) or not isinstance(encoded_text, tuple):
+        raise ValueError
+
+    end = storage.get_id('<END>')
+    upper_condition = False
+    sentence_decoded = []
+    text = []
+    for word_id in encoded_text:
+        if word_id == end:
+            upper_condition = True
+            text.append(" ".join(sentence_decoded))
+            sentence_decoded = []
+        else:
+            word = storage.get_word(word_id)
+            if upper_condition:
```

Comment:
This never returns to the `False` so that not only the first word of a sentence has an upper first letter but all other words too.
FInd a place to set this variable to the `False`.